File: internal/retrieval/internal/engine/metadata.go
package engine

import (
	"fmt"
	"sort"
	"strings"

	"github.com/danieljustus/symaira-desktop/internal/retrieval/internal/parser"
	"github.com/danieljustus/symaira-desktop/internal/vault"
)

const (
	searchMetadataStart           = "__SYMDESK_SEARCH_METADATA_START__"
	searchMetadataEnd             = "__SYMDESK_SEARCH_METADATA_END__"
	metadataTitleTagBoost float32 = 0.1
	metadataFieldBoost    float32 = 0.02
)

// SearchMetadata is the selected, already-parsed metadata included in the
// hybrid index representation. Fields are sorted into a canonical order when
// rendered, making repeatable indexing independent of map or caller order.
type SearchMetadata struct {
	Fields []SearchMetadataField
}

// SearchMetadataField is one scalar metadata value. Weight controls how many
// times the value is emitted in the representation; title and tags use the
// fixed higher weight chosen by the vault integration.
type SearchMetadataField struct {
	Name   string
	Value  string
	Weight int
}

// SearchMetadataFromVault maps the selected vault contract fields to the
// canonical search representation used by every local-file indexing path.
func SearchMetadataFromVault(doc *vault.Document) SearchMetadata {
	if doc == nil {
		return SearchMetadata{}
	}
	metadata := SearchMetadata{}
	add := func(name, value string, weight int) {
		if strings.TrimSpace(value) != "" {
			metadata.Fields = append(metadata.Fields, SearchMetadataField{Name: name, Value: value, Weight: weight})
		}
	}
	add("title", doc.Title, 3)
	add("tags", strings.Join(doc.Tags, " "), 3)
	add("aliases", strings.Join(doc.Aliases, " "), 2)
	add("created", doc.Created, 1)
	add("document_date", doc.DocumentDate, 1)
	add("person", doc.Person, 1)
	add("status", doc.Status, 1)
	add("due_date", doc.DueDate, 1)
	add("ocr_json_path", doc.OcrJSONPath, 1)
	add("simhash", doc.Simhash, 1)
	add("type", doc.Type, 1)
	if doc.ASN != nil {
		add("asn", fmt.Sprintf("%d", *doc.ASN), 1)
	}
	for _, key := range []string{"document_type", "correspondent", "source_path", "mime", "category", "ocr_engine", "archive_path", "imported_from", "import_run_id", "source_uri", "download_uri", "ingested_at", "sha256", "meeting_id", "notebook_id", "base_id"} {
		if value, ok := doc.Frontmatter[key]; ok {
			add(key, fmt.Sprint(value), 1)
		}
	}
	if value, ok := doc.Frontmatter["confidence"]; ok {
		add("confidence", fmt.Sprint(value), 1)
	}
	return metadata
}

// metadataSections prepends a synthetic metadata section while retaining all
// source sections and their existing location anchors. Synthetic chunks do not
// receive source character spans because their text is not present in the file.
func prependSearchMetadata(sections []parser.Section, metadata SearchMetadata) []parser.Section {
	text := formatSearchMetadata(metadata)
	if text == "" {
		return sections
	}
	result := make([]parser.Section, 0, len(sections)+1)
	result = append(result, parser.Section{
		Text:      text,
		Anchor:    parser.Anchor{Kind: "section", Value: "metadata"},
		Synthetic: true,
	})
	return append(result, sections...)
}

func formatSearchMetadata(metadata SearchMetadata) string {
	fields := append([]SearchMetadataField(nil), metadata.Fields...)
	sort.SliceStable(fields, func(i, j int) bool {
		left, right := metadataFieldRank(fields[i].Name), metadataFieldRank(fields[j].Name)
		if left != right {
			return left < right
		}
		if fields[i].Name != fields[j].Name {
			return fields[i].Name < fields[j].Name
		}
		return fields[i].Value < fields[j].Value
	})

	var b strings.Builder
	for _, field := range fields {
		name := strings.TrimSpace(field.Name)
		value := strings.Join(strings.Fields(field.Value), " ")
		if name == "" || value == "" {
			continue
		}
		weight := field.Weight
		if weight <= 0 {
			weight = 1
		}
		if weight > 4 {
			weight = 4
		}
		for i := 0; i < weight; i++ {
			fmt.Fprintf(&b, "%s: %s\n", name, value)
		}
	}
	if b.Len() == 0 {
		return ""
	}
	return searchMetadataStart + "\n" + b.String() + searchMetadataEnd
}

func metadataFieldRank(name string) int {
	switch name {
	case "title":
		return 0
	case "tags":
		return 1
	case "aliases":
		return 2
	case "created":
		return 3
	case "document_date":
		return 4
	case "due_date":
		return 5
	case "type":
		return 6
	case "status":
		return 7
	case "person":
		return 8
	case "correspondent":
		return 9
	case "asn":
		return 10
	default:
		return 100
	}
}

// MetadataMatches returns metadata field names whose values contain one of the
// free-text query terms. The order is the representation order and duplicates
// are removed. It is intentionally derived from the indexed text so old index
// rows remain readable and no database schema migration is needed.
func MetadataMatches(query, content string) []string {
	start := strings.Index(content, searchMetadataStart)
	if start < 0 {
		return nil
	}
	start += len(searchMetadataStart)
	end := strings.Index(content[start:], searchMetadataEnd)
	if end < 0 {
		return nil
	}
	metadataText := content[start : start+end]
	terms := strings.Fields(strings.ToLower(query))
	if len(terms) == 0 {
		return nil
	}

	seen := make(map[string]struct{})
	var matches []string
	for _, line := range strings.Split(metadataText, "\n") {
		name, value, ok := strings.Cut(line, ":")
		if !ok {
			continue
		}
		name = strings.TrimSpace(name)
		value = strings.ToLower(strings.TrimSpace(value))
		if name == "" || value == "" {
			continue
		}
		for _, term := range terms {
			term = strings.Trim(term, "\"'")
			if term != "" && strings.Contains(value, term) {
				if _, exists := seen[name]; !exists {
					seen[name] = struct{}{}
					matches = append(matches, name)
				}
				break
			}
		}
	}
	return matches
}

func metadataBoost(fields []string) float32 {
	for _, field := range fields {
		if field == "title" || field == "tags" {
			return metadataTitleTagBoost
		}
	}
	if len(fields) > 0 {
		return metadataFieldBoost
	}
	return 0
}

func StripSearchMetadata(content string) string {
	start := strings.Index(content, searchMetadataStart)
	if start < 0 {
		return content
	}
	end := strings.Index(content[start+len(searchMetadataStart):], searchMetadataEnd)
	if end < 0 {
		return strings.TrimSpace(content[:start])
	}
	end += start + len(searchMetadataStart) + len(searchMetadataEnd)
	return strings.TrimSpace(content[:start] + "\n" + content[end:])
}

File: crates/symdesk-vault/src/metadata.rs
#![deny(unsafe_code)]

use std::collections::HashSet;

use crate::Document;

const SEARCH_METADATA_START: &str = "__SYMDESK_SEARCH_METADATA_START__";
const SEARCH_METADATA_END: &str = "__SYMDESK_SEARCH_METADATA_END__";

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SearchMetadataField {
    pub name: String,
    pub value: String,
    pub weight: i32,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct SearchMetadata {
    pub fields: Vec<SearchMetadataField>,
}

#[must_use]
pub fn search_metadata_from_document(document: &Document) -> SearchMetadata {
    let mut fields = Vec::new();
    let mut add = |name: &str, value: String, weight| {
        if !value.trim().is_empty() {
            fields.push(SearchMetadataField {
                name: name.to_owned(),
                value,
                weight,
            });
        }
    };
    add("title", document.title.clone(), 3);
    add("tags", document.tags.join(" "), 3);
    add("aliases", document.aliases.join(" "), 2);
    add("created", document.created.clone(), 1);
    add("document_date", document.document_date.clone(), 1);
    add("person", document.person.clone(), 1);
    add("status", document.status.clone(), 1);
    add("due_date", document.due_date.clone(), 1);
    add("ocr_json_path", document.ocr_json_path.clone(), 1);
    add("simhash", document.simhash.clone(), 1);
    add("type", document.document_type.clone(), 1);
    if let Some(asn) = document.asn {
        add("asn", asn.to_string(), 1);
    }
    for key in [
        "document_type",
        "correspondent",
        "source_path",
        "mime",
        "category",
        "ocr_engine",
        "archive_path",
        "imported_from",
        "import_run_id",
        "source_uri",
        "download_uri",
        "ingested_at",
        "sha256",
        "meeting_id",
        "notebook_id",
        "base_id",
    ] {
        if let Some(value) = document.frontmatter.get(key) {
            add(key, value.to_string(), 1);
        }
    }
    if let Some(value) = document.frontmatter.get("confidence") {
        add("confidence", value.to_string(), 1);
    }
    SearchMetadata { fields }
}

#[must_use]
pub fn format_search_metadata(metadata: &SearchMetadata) -> String {
    let mut fields = metadata.fields.clone();
    fields.sort_by(|left, right| {
        metadata_field_rank(&left.name)
            .cmp(&metadata_field_rank(&right.name))
            .then_with(|| left.name.cmp(&right.name))
            .then_with(|| left.value.cmp(&right.value))
    });
    let mut lines = String::new();
    for field in fields {
        let name = field.name.trim();
        let value = field.value.split_whitespace().collect::<Vec<_>>().join(" ");
        if name.is_empty() || value.is_empty() {
            continue;
        }
        let weight = field.weight.clamp(1, 4);
        for _ in 0..weight {
            lines.push_str(name);
            lines.push_str(": ");
            lines.push_str(&value);
            lines.push('\n');
        }
    }
    if lines.is_empty() {
        String::new()
    } else {
        format!("{SEARCH_METADATA_START}\n{lines}{SEARCH_METADATA_END}")
    }
}

#[must_use]
pub fn metadata_matches(query: &str, content: &str) -> Vec<String> {
    let Some(start) = content.find(SEARCH_METADATA_START) else {
        return Vec::new();
    };
    let metadata_start = start + SEARCH_METADATA_START.len();
    let Some(relative_end) = content[metadata_start..].find(SEARCH_METADATA_END) else {
        return Vec::new();
    };
    let metadata = &content[metadata_start..metadata_start + relative_end];
    let terms: Vec<String> = query
        .to_lowercase()
        .split_whitespace()
        .map(|term| term.trim_matches(['"', '\'']).to_owned())
        .filter(|term| !term.is_empty())
        .collect();
    if terms.is_empty() {
        return Vec::new();
    }
    let mut seen = HashSet::new();
    let mut matches = Vec::new();
    for line in metadata.lines() {
        let Some((name, value)) = line.split_once(':') else {
            continue;
        };
        let name = name.trim();
        let value = value.trim().to_lowercase();
        if name.is_empty() || value.is_empty() {
            continue;
        }
        if terms.iter().any(|term| value.contains(term)) && seen.insert(name.to_owned()) {
            matches.push(name.to_owned());
        }
    }
    matches
}

#[must_use]
pub fn strip_search_metadata(content: &str) -> String {
    let Some(start) = content.find(SEARCH_METADATA_START) else {
        return content.to_owned();
    };
    let after_start = start + SEARCH_METADATA_START.len();
    let Some(relative_end) = content[after_start..].find(SEARCH_METADATA_END) else {
        return content[..start].trim().to_owned();
    };
    let end = after_start + relative_end + SEARCH_METADATA_END.len();
    format!("{}\n{}", &content[..start], &content[end..])
        .trim()
        .to_owned()
}

fn metadata_field_rank(name: &str) -> u8 {
    match name {
        "title" => 0,
        "tags" => 1,
        "aliases" => 2,
        "created" => 3,
        "document_date" => 4,
        "due_date" => 5,
        "type" => 6,
        "status" => 7,
        "person" => 8,
        "correspondent" => 9,
        "asn" => 10,
        _ => 100,
    }
}
